sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
     }
   }
 
-  def incRefCount(cb: ColumnarBatch): ColumnarBatch = {
-    GpuColumnVector.extractBases(cb).foreach(_.incRefCount())
-    cb
+  def extractTopLevelAttributes(
+      exprs: Seq[Expression],
+      includeAlias: Boolean): Seq[Option[Attribute]] =
+    exprs.map {
+      case a: AttributeReference => Some(a.toAttribute)
+      case GpuAlias(a: AttributeReference, _) if includeAlias => Some(a.toAttribute)
+      case _ => None
+    }
+
+  /**
+   * Filter rows from the batch where all of the keys are null.
+   */
+  def filterNulls(cb: ColumnarBatch, boundKeys: Seq[Expression]): ColumnarBatch = {
+    var mask: ai.rapids.cudf.ColumnVector = null
+    try {
+      withResource(GpuProjectExec.project(cb, boundKeys)) { keys =>
+        val keyColumns = GpuColumnVector.extractBases(keys)
+        // to remove a row all of the key columns must be null for that row
+        // If there is even one key column with no nulls in it, don't filter anything
+        // we do this by leaving mask as null
+        if (keyColumns.forall(_.hasNulls)) {
+          keyColumns.foreach { column =>
+            withResource(column.isNull) { nn =>
+              if (mask == null) {
+                mask = nn.incRefCount()
+              } else {
+                mask = withResource(mask) { _ =>
+                  mask.and(nn)
+                }
+              }
+            }
+          }
+        }
+      }
+
+      if (mask == null) {
+        // There was nothing to filter.
+        GpuColumnVector.incRefCounts(cb)
+      } else {
+        val colTypes = GpuColumnVector.extractTypes(cb)
+        withResource(GpuColumnVector.from(cb)) { tbl =>
+          withResource(tbl.filter(mask)) { filtered =>
+            GpuColumnVector.from(filtered, colTypes)
+          }
+        }
+      }
+    } finally {
+      if (mask != null) {
+        mask.close()
+        mask = null
+      }
+    }
+  }
+}
+
+/**
+ * Generic trait for all join gather instances.
+ * All instances should be spillable.
+ * The life cycle of this assumes that when it is created that the data and
+ * gather maps will be used shortly.
+ * If you are not going to use these for a while, like when returning from an iterator,
+ * then allowSpilling should be called so that the cached data is released and spilling
+ * can be allowed.  If you need/want to use the data again, just start using it, and it
+ * will be cached yet again until allowSpilling is called.
+ * When you are completely done with this object call close on it.
+ */
+trait JoinGatherer extends AutoCloseable with Arm {
```

Comment:
Would it be better to move this to its own file? I would like to use it with the broadcast nested loop join/cross join code too.

Reply:
+1